File: src/main/java/com/juanoff/types/UserType.java
package com.juanoff.types;

import java.io.InputStreamReader;

public interface UserType {
    String typeName();

    Object create();

    Object clone(Object obj);

    Object readValue(InputStreamReader in) throws Exception;

    Object parseValue(String ss) throws Exception;

    Comparator getTypeComparator();

    String serialize(Object object);

    Object deserialize(String s) throws Exception;
}

File: src/main/kotlin/com/juanoff/kotlin/types/UserType.kt
package com.juanoff.kotlin.types

import java.io.InputStreamReader

interface UserType {
    fun typeName(): String

    fun create(): Any

    fun clone(obj: Any): Any

    fun readValue(reader: InputStreamReader): Any

    fun parseValue(ss: String): Any

    fun getTypeComparator(): Comparator

    fun serialize(obj: Any): String

    fun deserialize(s: String): Any
}